module/zfs/zfs_znode.c

```diff
@@ -1049,6 +1075,11 @@ zfs_xvattr_set(znode_t *zp, xvattr_t *xvap, dmu_tx_t *tx)
 		    zp->z_pflags, tx);
 		XVA_SET_RTN(xvap, XAT_SPARSE);
 	}
+	if (XVA_ISSET_REQ(xvap, XAT_PROJINHERIT)) {
+		ZFS_ATTR_SET(zp, ZFS_PROJINHERIT, xoap->xoa_projinherit,
+		    zp->z_pflags, tx);
```

Comment:
@ahrens As I explained in former comment, I agree that there will be more work for tree quota. But I do NOT think it will be the reason to drop the PROJINHERIT flag. Because we still need to offer some mechanism to the user/application to set project ID automatically (for convenient usage). As you know, when a file is created, its UID/GID attribute can be set based on the uid/gid information associated with the task_struct; unfortunately, there is no project ID information in the task_struct. Then inheriting the project ID from its parent directory become the most intuitive and simple solution.

About the tree quota, I think that we can keep the PROJINHERIT flag there, it offers some basic framework for tree quota. For example, I can create a new directory, and set permission that only myself can modify the directory, then I can use the incomplete tree quota to calculate the space used by all the objects under such directory, that is much faster than the normal "du" for very large directory. So the user who follows some rules still can use such PROJINHERIT flag for some tree quota purpose. But of course, we need continue work to enhance the tree quota in the future.